### DISCUSSION TITLE
What is the minimum required to build a plugin?

### DISCUSSION THREAD

**Discussion by Ddiidev** [Flow-Launcher/Flow.Launcher#2087] [Q&A]
Hello, I'm trying to build a plugin in **Vlang**, but I don't quite understand how to do it.
I didn't understand how to **install a plugin manually** without the need for the store.

**"ExecutablePath"** I can pass a ".exe"? what does main(or similar) expect?

Can someone get me an implementation example in **Golang/rust/lua** etc... that doesn't need the FlowLauncher sdk, so that I can implement it in an equivalent way in Vlang.

**Comment by Garulf** [comment]
Hello @ldedev,

To answer your first question Flow Launcher will search every folder in its `plugins` directory that includes a manfiest file named `plugin.json`. You can manually place your plugin there and restart Flow Launcher to have it load your plugin. You do not need to restart Flow Launcher if you code changes, only if your manifest changes. You can test changes in your plugin very easily this way.

> "ExecutablePath" I can pass a ".exe"? what does main(or similar) expect?

The short answer is yes.

As long as the program can receive arguments and communicate to `stdout`. Flow Launcher uses JsonRPC to send and receive data.

> Can someone get me an implementation example in Golang/rust/lua etc... that doesn't need the FlowLauncher sdk, so that I can implement it in an equivalent way in Vlang.

Of those languages i am only familar with LUA really so I can provide a simple "class" that communicates with Flow Launcher like so:

(I just whipped this up as an example, code could be better?)
```lua
local json = require("json")

local FlowLauncher = {}

function FlowLauncher:new()
  local newObj = {}
  self.__index = self
  return setmetatable(newObj, self)
end

function FlowLauncher:run()
    self.rpc_request = json.decode(arg[1])

    local method = self.rpc_request["method"]
    local parameters = self.rpc_request["parameters"]
    local results = self[method](self, unpack(parameters))
    
    output = {
        ["results"] = results
    }
    print(json.encode(output))


end


function FlowLauncher:query(param)
  -- This method is ran when the user enters your plugin keyword on Flow Launcher
  return {}
end

function FlowLauncher:context_menu(data)
  -- optional context menu entries for a result
  return {}
end
```

Plugin example:

input: `'{"method": \query", "parameters": ["test", "test1"]}'`
```lua
local myPlugin = FlowLauncher:new()

function myPlugin:query(param)
    local result = {
        {
            Title = "Hello World",
            SubTitle = "This is the first plugin",
            IcoPath = "Images\\ico.png",
        }
    }
    return result
end

myPlugin:run()
```
Output:
`{"results":[{"SubTitle":"This is the first plugin","Title":"Hello World","IcoPath":"Images\\ico.png"}]}`

**Reply by Ddiidev** [reply]
Hello, I tried to do exactly as you explained, but it started not working when the flowlauncher found my plugin.

See if the structure is correct...

![image](https://user-images.githubusercontent.com/7676415/234931863-bbe928bf-c40b-410f-b81c-f9ffb30dc78f.png)

plugin.json
```json
{
    "ID": "27afe2db-b114-4783-83b7-bf9f47288d4a",
    "ActionKeyword": "v",
    "Name": "Teste",
    "Description": "Vlang Teste",
    "Author": "André Luiz",
    "Version": "1.0.0",
    "Language": "Vlang",
    "Website": "",
    "IcoPath": "src/assets/33x33.png",
    "ExecuteFileName": "src/main.exe"
}
```

![image](https://user-images.githubusercontent.com/7676415/234932267-396a87b9-0418-4307-926f-132c544b79d8.png)


Do you know what I can do?

**Reply by Garulf** [reply]
Structure and plugin manifest look correct. 

Have you restarted Flow Launcher so it can load new plugins?

If so please share your logs maybe we can see what happened from an error message.

**Reply by Ddiidev** [reply]
Sorry for the delay in responding...
If you can still help me, thank you!

Well I'll send you a zipped file of how I configured it on my machine, see if it's ok...

[Teste-1.0.0.zip](https://github.com/Flow-Launcher/Flow.Launcher/files/11582305/Teste-1.0.0.zip)

When running the src/main.exe file
it should just create a text file called 'text.txt'
